lib/sourceror/zipper.ex

```diff
@@ -494,4 +494,27 @@ defmodule Sourceror.Zipper do
 
   defp do_append_child(list, child) when is_list(list), do: list ++ [child]
   defp do_append_child({left, right}, child), do: {:{}, [], [left, right, child]}
+
+  defimpl Inspect do
+    import Inspect.Algebra
+
+    def inspect(zipper, opts) do
+      opts.custom_options
+      |> Keyword.get(:zipper, :as_ast)
+      |> inspect(zipper, opts)
+    end
+
+    defp inspect(:as_ast, zipper, opts) do
+      concat(["#Zipper<", Inspect.List.keyword({:node, zipper.node}, opts), ">"])
+    end
+
+    defp inspect(:as_code, zipper, opts) do
+      code = Sourceror.to_algebra(zipper.node, Map.to_list(opts))
+      nest(concat(["#Zipper<node:", code, ">"]), 1) |> format(opts.width)
```

Comment:
An early comment:
We probably want to use something a little more coplex here to unsure better fit, like Nx does to print the tensors
https://github.com/elixir-nx/nx/blob/0af36df37f6b7582e242a9aa19748fb6a5e12dcd/nx/lib/nx/tensor.ex#L241-L248

Reply:
Yeah, for sure -- I had to stop work on this early and was just committing what I had :)

Reply:
I’d like to revisit `Inspect` later today.

I’ve been back and forth a lot on how to clearly surface useful information about “where you are” in the Zipper when displaying `:as_ast` or `:as_code`. The simplest thing is to just only show the current node, like

```
# :as_ast
#Sourceror.Zipper<
  {:def, [], [...]}
>

# :as_code
#Sourceror.Zipper<
  def foo do
    :bar
  end
>
```

But there are other potentially useful things to indicate that would normally be visible by looking at the path a bit:

- do we have ancestors/are we at the root?
- do we have siblings?

Spitballing a bit:

```
# :as_ast
#Sourceror.Zipper<
  #root
  {:def, [], [...]}
>

#Sourceror.Zipper<
  #...
  {:def, [], [...]}
  #...
>

# :as_code
#Sourceror.Zipper<
  #root
  def foo do
    :bar
  end
>

# all potentially valid:

#Sourceror.Zipper<
  #...
  def foo do
    :bar
  end
  #...
>

#Sourceror.Zipper<
  def foo do
    :bar
  end
  #...
>

#Sourceror.Zipper<
  #...
  def foo do
    :bar
  end
>

#Sourceror.Zipper<
  def foo do
    :bar
  end
>
```

Is it obvious what the `#root` and `#...` indicate here? I was writing out an explanation, but realized that a better "test" would be whether it is self-evident. :)

Reply:
I went ahead and implemented it this way for now -- curious for your feedback when you have time!

Reply:
One issue I found with `:as_code` -- there are situations where what's being formatted is a section of the AST that you can get to via the Zipper but isn't really valid AST, so the formatting you get is a bit weird/unexpected:

```elixir
iex(22)> IEx.configure(inspect: [custom_options: [zipper: :as_code]])
:ok
iex(23)> "Foo.Bar.baz()" |> Code.string_to_quoted!() |> Z.zip()
#Sourceror.Zipper<
  #root
  Foo.Bar.baz()
>
iex(24)> "Foo.Bar.baz()" |> Code.string_to_quoted!() |> Z.zip() |> Z.next()
#Sourceror.Zipper<
  Foo.Bar . :baz
>
```

Reply:
`:as_code` can also be a bit weird when you're dealing with Sourceror AST, which wraps literals in `:__block__` tuples:

```
iex(5)> "1 + 2" |> Sourceror.parse_string!() |> Z.zip()
#Sourceror.Zipper<
  #root
  1 + 2
>
iex(6)> "1 + 2" |> Sourceror.parse_string!() |> Z.zip() |> Z.next()
#Sourceror.Zipper<
  1
  #...
>
iex(7)> "1 + 2" |> Sourceror.parse_string!() |> Z.zip() |> Z.next() |> Z.next()
#Sourceror.Zipper<
  1
>
```

Reply:
I think `#root` and `#...` are ok, but we might want to use a different color with `IO.ANSI` to make it clear it's a zipper formatting thing, and not an actual comment in the original code. We can use a greyish color for this

I wouldn't worry too much about a perfect representation of invalid ast; I think there are many disparate ways in which you can reach them, and this is a debugging utility so it shouldn't be too probolematic

Reply:
Good idea! I looked into this initially but there wasn't a default "faint/subdued" color in `:syntax_colors`, which are used with `Inspect.Algebra.color/3`. We can add an arbitrary one though:

```elixir
opts = Map.update!(opts, :syntax_colors, &Keyword.put_new(&1, :zipper_internal, :light_black))

...

color("#root", :zipper_internal, opts)
```

Reply:
![image](https://github.com/doorgan/sourceror/assets/503938/951988f5-0c40-42c5-9b08-ebf9663c96eb)